Katello/katello
# Fixes #33762 - Add toggle group for All/Installable  errata

### What are the changes introduced in this pull request?
On the new host errata page
- Added toggle group with options for All / Installable
- The "All" option is what was previously known as "Applicable" errata
- Added an Installable column to indicate if the errata is installable or not "No"
- Action menu items (apply etc.) should be disabled for applicable (but not installable) errata
- Also ability to handle "unselectable" rows when using BulkSelect and SelectionSet hooks

### What are the testing steps for this pull request?

1. Create and Sync this repo -> https://partha.fedorapeople.org/test-repos/multi-errata/
2. Create another repo in the same product. Leave the feed empty
3. Create  a Dev lifecycle environment
4. Create a content view with the repos created in steps 1 and 2
5. Publish and Promote it to dev
6. Create an activation that points dev/cv. Make sure the subscriptions are turned on
7. Register a rhel7/centos7 with that activation key
8. On client `yum install armadillo-0.1-1.noarch`
9. Also `rpm -ivh https://jlsherrill.fedorapeople.org/fake-repos/needed-errata/walrus-0.71-1.noarch.rpm https://jlsherrill.fedorapeople.org/fake-repos/needed-errata/whale-0.2-1.noarch.rpm https://jlsherrill.fedorapeople.org/fake-repos/needed-errata/shark-0.1-1.noarch.rpm https://jlsherrill.fedorapeople.org/fake-repos/needed-errata/stork-0.12-2.noarch.rpm`
10. Now edit the repo in step 2 and add the feed url -> https://jlsherrill.fedorapeople.org/fake-repos/needed-errata/
11. Sync this repo
12. Now you should have 38 Applicable Errata and 37 installable errata
13. Go to Hosts -> New Host Details -> Content -> Errata tab
14. Notice the new toggle group.
15. Choose all. Notice that the first check box must be disabled. You should be able hover on 'X' in the installable column and and notice a tooltip that explains why that is the case
16. Try all the select options , add search etc. 
17. Report any inconsistencies
18.
![toggle](https://user-images.githubusercontent.com/1069779/138371692-33318c43-b424-486a-ac9f-b3ef4bc63651.png)


### Guidelines for Code Review ###

#### Backend Changes #####

- Added a new parameter to host_errata_controller to accept a `restrict_applicable` flag. Setting this to true will automatically include applicable errata in the index call
- Added a `selectable` attribute to `_metadata.json.rabl`. This will hold the count of selectable items in main list. We need this information for select all to correctly display how many items are going to be selected.
-  Added mechanism  to set selectable in `v2/api_controller`.  Individual controllers can override the "total_selectable" method and return count of selectable items. (By default selectable === total)
- 

#### Front End Changes ####
* In *TableHooks* ->  UseSelectionSet and useBulkSelect rewired a little now to accept an `isSelectable` function. Based on this those hooks will calculate things like `selectedCount`, `areAllRowsSelected` etc.
* In *TableWrapper* a new place holder was added for toggleGroups.
* In *ErrataTab* new toggleGroup for All/Installable, a new State to handle if its the All toggle or the installable toggle.
* Also includes some tool tips and other information

- Added Tests
